solr/solrj/src/java/org/apache/solr/common/util/XML.java

```diff
@@ -101,82 +102,56 @@ public final static void writeXML(Writer out, String tag, String val) throws IOE
+    writeXML(out, tag, valWritable, null, attrs);
+  }
+
+  private final static void writeXML(Writer out, String tag, Writable valWritable,
+                                     Escapable<String, Writer> attrWriteFunc, Object... attrs) throws IOException {
     out.write('<');
     out.write(tag);
     for (int i=0; i<attrs.length; i++) {
       out.write(' ');
       out.write(attrs[i++].toString());
       out.write('=');
       out.write('"');
-      out.write(attrs[i].toString());
+      if(attrWriteFunc != null)
+        attrWriteFunc.escape(attrs[i].toString(), out);
+      else
+        out.write(attrs[i].toString());
       out.write('"');
     }
-    if (val == null) {
+    if (valWritable == null) {
       out.write('/');
       out.write('>');
     } else {
       out.write('>');
-      out.write(val);
+      valWritable.write(out);
       out.write('<');
       out.write('/');
       out.write(tag);
       out.write('>');
     }
   }
 
-  /** escapes character data in val */
+  /** escapes character data in val and attributes */
   public final static void writeXML(Writer out, String tag, String val, Object... attrs) throws IOException {
-    out.write('<');
-    out.write(tag);
-    for (int i=0; i<attrs.length; i++) {
-      out.write(' ');
-      out.write(attrs[i++].toString());
-      out.write('=');
-      out.write('"');
-      escapeAttributeValue(attrs[i].toString(), out);
-      out.write('"');
-    }
-    if (val == null) {
-      out.write('/');
-      out.write('>');
-    } else {
-      out.write('>');
-      escapeCharData(val,out);
-      out.write('<');
-      out.write('/');
-      out.write(tag);
-      out.write('>');
-    }
+    final Writable writable = val!= null? (writer1) -> XML.escapeCharData(val, writer1): null;
+    writeXML(out, tag, writable, XML::escapeAttributeValue, attrs);
   }
 
-  /** escapes character data in val */
+  /** escapes character data in val and attributes */
   public static void writeXML(Writer out, String tag, String val, Map<String, String> attrs) throws IOException {
-    out.write('<');
-    out.write(tag);
-    for (Map.Entry<String, String> entry : attrs.entrySet()) {
-      out.write(' ');
-      out.write(entry.getKey());
-      out.write('=');
-      out.write('"');
-      escapeAttributeValue(entry.getValue(), out);
-      out.write('"');
-    }
-    if (val == null) {
-      out.write('/');
-      out.write('>');
-    } else {
-      out.write('>');
-      escapeCharData(val,out);
-      out.write('<');
-      out.write('/');
-      out.write(tag);
-      out.write('>');
-    }
+    writeXML(out, tag, val, attrs.entrySet().stream().flatMap((entry) -> Stream.of(entry.getKey(), entry.getValue())).toArray());
   }
 
-  private static void escape(char [] chars, int offset, int length, Writer out, String [] escapes) throws IOException{
+  public static void escape(char [] chars, int offset, int length, Writer out, String [] escapes) throws IOException{
```

Comment:
Why make this public?  It seems rather internal: it isn't even documented; you have to really read the code to know what it does and what "escapes" is.

Reply:
Sorry forgot to revert this change. I was playing around with the code. Will switch it back to private.